DMTCP_TMPDIR is base dir only, dmtcp-USER@HOST added internally [CORRECTED]

gc00 opened the pull request:
This pull request has two commits.  They correspond to the first two items in Issue #48.
1.  Previously, even if the user does not set DMTCP_TMPDIR, DMTCP would set it anyway.  DMTCP should not be creating additional environment variables visible to the user program.  The extra environment variable has now been removed.
2.  DMTCP_TMPDIR (and the value for the --tmpdir flag) are now considered base directories only.  Internally, DMTCP appends "dmtcp-USER@HOST" as a subdirectory to the base directory in order to create the true tmpdir.  This is hidden, and not documented to the user.  The user should not be writing files here.

karya0 commented on src/dmtcp_coordinator.cpp:
```diff
@@ -1796,7 +1799,7 @@ int main ( int argc, char** argv )
       shift; shift;
     }else if(argc>1 && (s == "-t" || s == "--tmpdir")){
-      setenv(ENV_VAR_TMPDIR, argv[1], 1);
+      tmpDir = Util::calcTmpDir(argv[1]);
```
We need to use the `tmpdir_arg` trick here as well (just like dmtcp_restart and dmtcp_launch below).

gc00 replied:
Do we need the tmpdir_arg trick?  This code is inside:  dmtcp_coordinator.cpp:main().
Because it's invoked at the top level function, it should be enough to set the local variable tmpDir, and then pass that value into other functions.

karya0 replied:
If '--tmpdir' argument is not passed in then, the value of `tmpdir` would be "" and that would be a problem :-). Bottom line is that we need to call calcTmpDir() even if no --tmpdir flag was specified.

gc00 replied:
I fixed this now, but I don't know if github saved my comment.  Thanks for the catch.

karya0 commented:
PS: I am still not quite okay with the help string. Will try to come up with a suggestion by tomorrow, if I can.

gc00 commented:
I've fixed the --tmpdir issue ("tmpdir_arg trick").  I've also now updated the help on it in dmtcp_restart.cpp, to more accurately reflect what's happening.

karya0 commented:
The code looks good. However, I have a concern about the following sentence in the help message:
 "Can preempt previous tmpdir of coord or earlier launch.\n"

Apparently, we need to have the same tmpdir for all processes on each individual host because the pid plugin and others rely on files in tmpdir for publish/subscribe.  So if there are two processes A and B with two different tmpdirs tmpA and tmpB respectively, they won't be able to share the pids or participate in leader election for files, etc.

Thus I think we should make it clear in the help message that the tmpdirs for dmtcp_launch/dmtcp_restart SHOULD be the same for all processes on a given node.

gc00 commented:
Good point.  I'll change the help message to:
  "      Can preempt previous tmpdir of coord.  (Behavior is undefined\n"
  "        if two launched processes specify different tmpdirs.)\n"
and then check it in.

On Tue, Mar 31, 2015 at 10:49:03AM -0700, Kapil Arya wrote:
> The code looks good. However, I have a concern about the following sentence in the help message:
>  "Can preempt previous tmpdir of coord or earlier launch.\n"
> 
> Apparently, we need to have the same tmpdir for all processes on each individual host because the pid plugin and others rely on files in tmpdir for publish/subscribe.  So if there are two processes A and B with two different tmpdirs tmpA and tmpB respectively, they won't be able to share the pids or participate in leader election for files, etc.
> 
> Thus I think we should make it clear in the help message that the tmpdirs for dmtcp_launch/dmtcp_restart SHOULD be the same for all processes on a given node.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/dmtcp/dmtcp/pull/50#issuecomment-88186132

karya0 commented:
Do we need to say anything about coord tmpdir in the help messages for dmtcp_launch and restart? I am just thinking whether or not it is going to be confusing for the user.

gc00 commented:
Suppose I also put the "coord tmpdir" inside the parentheses:
"        Directory to store temporary files (default: $TMDPIR or /tmp)\n"
"        (Can preempt previous tmpdir of coord.  Behavior is undefined\n"
"         if two launched processes specify different tmpdirs.)\n"
But if you still prefer removing the sentence "Can preempt ...", tell me,
and I'm find with removing it.

On Tue, Mar 31, 2015 at 12:43:41PM -0700, Kapil Arya wrote:
> Do we need to say anything about coord tmpdir in the help messages for dmtcp_launch and restart? I am just thinking whether or not it is going to be confusing for the user. 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/dmtcp/dmtcp/pull/50#issuecomment-88223140

karya0 commented:
What if we never specified a --tmpdir for the coordinator?

Alternatively, what if a --tmpdir for the coordinator but we do not specify a --tmpdir for launch/restart. In this case, doesn't this somewhat gives the wrong message that coord tmpdir is inherited by launch/restart?

This is less apparent in single node computations, but in distributed computations, this might be a bit confusing. 

Having said all this, what if we change the wording to something like:
" Directory to store temporary files for dmtcp_launch ONLY (default: ...)\n
  Behavior is undefined if two ...."

karya0 commented:
I didn't read the whole message earlier. Yes, I would like to remove the "Can preempt .." sentence :-).

gc00 commented:
So, for now, I'll remove the "Can preempt ..." sentence.  The sentence:
   (Behavior is undefined if two launched processes specify different tmpdirs.)\n"
will remain.
Are there any other changes we should make before I commit this?

Best,
- Gene

On Tue, Mar 31, 2015 at 12:56:36PM -0700, Kapil Arya wrote:
> I didn't read the whole message earlier. Yes, I would like to remove the "Can preempt .." sentence :-).
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/dmtcp/dmtcp/pull/50#issuecomment-88227129

karya0 commented:
Sounds good. Let's commit it.

gc00 commented:
This has now been pushed into origin/master at:
  c458e6a43dbc30167927798287e674b86bac2e90
  b4d3b1b29b55ea9e05651fe6a511194b8bf563ee
